src/physics/material/MaterialParams.cc

```diff
@@ -198,39 +210,57 @@ MaterialParams::extend_elcomponents(const MaterialInput& inp)
 {
-    REQUIRE(inp.number_density >= 0);
-    REQUIRE((inp.number_density == 0) == inp.elements.empty());
     REQUIRE(inp.temperature > 0 || inp.number_density == 0);
 
-    auto insertion = matname_to_id_.insert(
-        {inp.name, MaterialDefId(host_materials_.size())});
-    CHECK(insertion.second);
+    auto iter = matname_to_id_.find(inp.name);
+    // If material exists: log warning
+    if (iter != matname_to_id_.end())
+    {
+        CELER_LOG(warning) << "Material " << inp.name
+                           << " already exists with id " << iter->second.get()
+                           << ". This new id (" << host_materials_.size()
+                           << ") will not be available.";
+    }
+    // else, add material to list
+    else
+    {
+        auto insertion = matname_to_id_.insert(
+            {inp.name, MaterialDefId(host_materials_.size())});
+    }
 
     MaterialDef result;
-
     // Copy basic properties
     result.number_density = inp.number_density;
     result.temperature    = inp.temperature;
     result.matter_state   = inp.matter_state;
-
-    // Set elemental fractions
-    result.elements = this->extend_elcomponents(inp);
-
-    // Renormalize component fractions and calculate weighted properties
-    real_type avg_amu   = 0;
-    real_type avg_z     = 0;
-    real_type rad_coeff = 0;
+    result.elements       = this->extend_elcomponents(inp);
+
+    /*!
+     * Calculate derived quantities: density, electron density and rad length
+     *
+     * NOTE: Electron density calculation may need to be updated for solids
+     */
+    result.electron_density  = 0;
+    real_type avg_molar_mass = 0;
+    real_type rad_coeff      = 0;
     for (const MatElementComponent& comp : result.elements)
     {
         CHECK(comp.element < host_elements_.size());
         const ElementDef& el = host_elements_[comp.element.get()];
-        avg_amu += comp.fraction * el.atomic_mass.value();
-        avg_z += comp.fraction * el.atomic_number;
+        // Calculate average molar mass
+        avg_molar_mass += comp.fraction * el.atomic_mass.value();
+        // Calculate element number density
+        real_type el_number_density = comp.fraction * inp.number_density;
+        // Add element electron density to the total electron density
+        result.electron_density += el_number_density * el.atomic_number;
+        // Calculate radiation coefficient
         rad_coeff += comp.fraction * el.mass_radiation_coeff;
     }
-    result.density = avg_amu * constants::atomic_mass * result.number_density;
-    result.electron_density = avg_z * result.number_density;
-    result.rad_length       = 1 / (rad_coeff * result.density);
+    // Calculate density
+    result.density = result.number_density * avg_molar_mass
+                     / constants::na_avogadro;
+    // Calculate radiation length from the radiation coefficient
+    result.rad_length = 1. / (rad_coeff * result.density);
```

Comment:
I think these changes can be reverted, right? The only difference is that you're now doing an extra multiply inside the loop (for the `el_number_density`) to avoid the single multiply at the end (`result.electron_density = avg_z * result.number_density`). If you think the nomenclature change (`avg_amu -> avg_molar_mass`) is more accurate you can keep that of course.

Reply:
I'll revert that. I wrote that explicitly while debugging (checking each element individually) and did not worry to revert it back, since this code runs once at initialization and seemed a bit more obvious (at least to me).

The `avg_molar_mass` seems more appropriate for 2 reasons: 
- Unit checking is clear in `result.density`, even though M and A are numerically equivalent.
- I opted for `N_A` instead of `amu_constant` while calculating `result.density` because `N_A` is now fixed, while the latter has an uncertainty attached to it.